compiler/test/dotty/tools/dotc/reporting/UrlExistenceChecker.scala

```diff
@@ -0,0 +1,103 @@
+package dotty.tools.dotc.reporting
+
+import java.net.{HttpURLConnection, SocketTimeoutException, URL, UnknownHostException}
+import java.util.concurrent.atomic.AtomicBoolean
+
+import org.junit.Assert.fail
+
+import scala.concurrent.{Await, Future}
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.util.control.NonFatal
+
+/** Checks URL existence by HTTP requests expecting OK responses (200).
+  *
+  * To not fail if no network is available all URL tests are cancelled if in case
+  * "https://www.github.com/" is not available.
+  *
+  * URL testing can be switched off via system property `linkChecking=false`.
+  */
+trait UrlExistenceChecker {
```

Comment:
The tests are far too complicated in my opinion. It should probably be a single simple test. Something like:
```scala
@Test
def testLinks = {
  def checkLink(link: DocumentationLink) = {
    // assert the link is valid
  }
  // keep list in sync with links in messages.scala
  List(link1, link2, link3, ...).foreach(checkLink)
}
```

Reply:
That's basically what's going on, except that:

- Everything is lifted into a Future and sequenced (makes all the network run in parallel!)
- There is no hand maintained list of links to check, this list is build by hooking into the existing error message tests and mutating `var checks`.

Reply:
I don't see a list of links anywhere, can you point me to the line?

I think in this case the logic is not unnecessary, we already have trouble reviewing the error message contributions, trading a bit of simplification on test code (which is nicely safe isolated in this file BTW) against additional complexity for contributors is not a clear win...

Reply:
https://github.com/ennru/dotty/blob/4b12de221b986913e2ac22a67a23441783be5ad0/compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala#L102-L108

There is no complexity for contributors. When they add a new error message, they add a new test. So if they add a new link, they dimply add a new entry to the list of links tested. I would argue that the current solution is actually more complex for contributors. It introduce a new magic method `assertMessageCountAndCheckLinks` which is not part of the documentation on error messages (blog, tutorial, issues).

Reply:
That's only the ones that are factored out between several error messages, they are all the inlined ones: https://github.com/ennru/dotty/blob/4b12de221b986913e2ac22a67a23441783be5ad0/compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala#L194. You see, we would have already missed like 10 URLs :P

I still think that having to manually maintain a list to all these links in tests is not a clear win over the implementation of this PR.

Maybe we could revert the name change from `assertMessageCountAndCheckLinks` back to `assertMessageCount`? This PR is already 6 months old, so lets stick to very concrete suggestions...

Reply:
The age of the PR should not be important for the decision to merge it.

The links could be introduced with much simpler code. My focus was on making the link checking as little annoying as possible. That's where the complexity comes in. 

No problem if the Dotty team decides to drop this.